physics_simulation.py

```diff
@@ -140,4 +160,91 @@ def get_stats(t: int):
             entity_stats = get_entity_stats(t, entity, stats)
             stats.append(entity_stats)
 
+    return get_connections(stats)
+
+
+# given two points to make a line, and a point
+# find the closest distance from the point and
+# the line. 
+# equation: https://en.wikipedia.org/wiki/Distance_from_a_point_to_a_line#Line_defined_by_two_points
+# algorithm: https://stackoverflow.com/questions/849211/shortest-distance-between-a-point-and-a-line-segment
+# (line_x1, line_y1), (line_x2, line_y2): 2 points making the line
+# (point_x, point_y): point
+def point_dist_to_line(line_x1: float, line_y1: float, line_x2: float, line_y2: float, point_x: float, point_y: float):
+    a = point_x - line_x1
+    b = point_y - line_y1
+    c = line_x2 - line_x1
+    d = line_y2 - line_y1
+
+    lenSq = c * c + d * d
+    param = -1
+
+    if lenSq != 0: # in case of 0 length line
+        dot = a * c + b * d
+        param = dot / lenSq
+
+    xx = 0
+    yy = 0
+
+    if param < 0:
+        xx = line_x1
+        yy = line_y1
+    elif param > 1:
+        xx = line_x2
+        yy = line_y2
+    else:
+        xx = line_x1 + param * c
+        yy = line_y1 + param * d
+
+    dx = point_x - xx
+    dy = point_y - yy
+
+    return math.sqrt(dx * dx + dy * dy)
+
+
+# get the connections between any entities
+# checking if an entity has a direct line of sight with the other entities
+# takes in a data structure stats which has all of the meta data for all of the 
+# entities, including:
+# "id": id of the entity
+# "orbital_radius": radius of the orbit (m)
+# "orb_id": id of the entity which this entity orbits
+# "period": orbital period (s)
+# "radius": radius of the planet (m)
+# "can_connect": entity able to connect to the interplanetary internet
+# "x": x coordinate in space (m)
+# "y": y coordinate in space (m)
+def get_connections(stats):
+    
+    # If anyone has a better idea that isnt O(n^3) I am listening
```

Comment:
Off the top of my head, it does seem like O(n^3) is an easy to understand and close-to-optimal solution. There are probably ways to bring it up to O(n^2log(n)) or O(n^2), but it's probably more trouble than it's worth.

Reply:
Yeah, its not really worth the effort when I cant see us going past the low double digits for number of entities. I think what we would have to do is to make it so that each node only checks a certain number of nodes which is in the neighborhood. That or some kind of actual discovery algorithm.